# (1) load_digits : 손글씨 분류

In [11]:
#모듈 import
from sklearn.datasets import load_digits 
from sklearn.model_selection import train_test_split  #데이터셋 분리 모듈
from sklearn.metrics import classification_report

#데이터 준비
digits = load_digits()
digits_data = digits.data 
digits_label = digits.target  #digits_label변수에 digits 데이터의 target저장

In [13]:
#Target Names출력
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [14]:
#데이터 Describe
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [18]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data,  #feature
                                                    digits_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성
                                                   

In [36]:
#decision tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360

0.8555555555555555


In [35]:
#random forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       0.93      1.00      0.96        37
           5       0.90      0.96      0.93        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       1.00      0.84      0.91        43
           9       0.94      0.94      0.94        32

    accuracy                           0.96       360
   macro avg       0.96      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360

0.9638888888888889


In [37]:
#svm
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360

0.9888888888888889


In [38]:
#sgd classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.98      1.00      0.99        43
           1       0.90      0.88      0.89        42
           2       1.00      1.00      1.00        40
           3       0.94      0.91      0.93        34
           4       0.97      0.97      0.97        37
           5       0.90      0.96      0.93        28
           6       0.93      0.93      0.93        28
           7       0.91      0.97      0.94        33
           8       0.97      0.81      0.89        43
           9       0.86      0.97      0.91        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360

0.9388888888888889


In [39]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인



              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360

0.9527777777777777


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#모델 평가하기
Decision tree : 85.5%
random forest : 96.3%
svm : 98.8%
sgd classifier : 93.8%
Logistic Regression : 95.2%

svm모델 사용시 accuracy가 98.8로 가장 높게 도출되었다.

# (2) load_wine : 와인 분류

In [42]:
#모듈 import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#데이터 준비
wine = load_wine()
wine_data = wine.data 
wine_label = wine.target

In [41]:
#Target Names 출력
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [43]:
#데이터 Describe
print(digits.DESCR)

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

In [44]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data,  #feature
                                                    wine_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성
                                                   

In [45]:
#decision tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.89      1.00      0.94        17
           2       1.00      0.83      0.91        12

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36

0.9444444444444444


In [46]:
#random forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        12

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36

1.0


In [47]:
#svm
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.86      0.86      0.86         7
           1       0.58      0.88      0.70        17
           2       0.33      0.08      0.13        12

    accuracy                           0.61        36
   macro avg       0.59      0.61      0.56        36
weighted avg       0.55      0.61      0.54        36

0.6111111111111112


In [48]:
#sgd classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.88      1.00      0.93         7
           1       0.92      0.71      0.80        17
           2       0.73      0.92      0.81        12

    accuracy                           0.83        36
   macro avg       0.84      0.87      0.85        36
weighted avg       0.85      0.83      0.83        36

0.8333333333333334


In [49]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인



              precision    recall  f1-score   support

           0       1.00      0.86      0.92         7
           1       0.94      1.00      0.97        17
           2       1.00      1.00      1.00        12

    accuracy                           0.97        36
   macro avg       0.98      0.95      0.96        36
weighted avg       0.97      0.97      0.97        36

0.9722222222222222


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#모델 평가하기
Decision tree : 94.4%
random forest : 100%
svm : 61.1%
sgd classifier : 83.3%
Logistic Regression : 97.2%

random forest모델 사용시 accuracy가 100으로 가장 높게 도출되었다.

---

# (3) load_breast_cancer : 유방암 여부 진단

In [51]:
#모듈 import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#데이터 준비
breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data 
breast_cancer_label = breast_cancer.target

In [52]:
#Target Names 출력
breast_cancer.target_names

array(['malignant', 'benign'], dtype='<U9')

In [53]:
#데이터 Describe
print(breast_cancer.DESCR)

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [54]:
#trian, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data,  #feature
                                                    breast_cancer_label,  #label
                                                    test_size=0.2,   #test dataset 크기
                                                    random_state=7)  #데이터 분리 시 랜덤성
                                                   

In [55]:
#decision tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114

0.9122807017543859


In [56]:
#random forest
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32) 
random_forest.fit(X_train, y_train) 
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      1.00      1.00        74

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114

1.0


In [57]:
#svm
from sklearn import svm

svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114

0.9035087719298246


In [58]:
#sgd classifier
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인

              precision    recall  f1-score   support

           0       1.00      0.55      0.71        40
           1       0.80      1.00      0.89        74

    accuracy                           0.84       114
   macro avg       0.90      0.78      0.80       114
weighted avg       0.87      0.84      0.83       114

0.8421052631578947


In [59]:
#Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

print(accuracy_score(y_test, y_pred)) #정확도 확인



              precision    recall  f1-score   support

           0       1.00      0.85      0.92        40
           1       0.93      1.00      0.96        74

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114

0.9473684210526315


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#모델 평가하기
(1) accuracy
Decision tree : 91.2%
random forest : 100%
svm : 90.3%
sgd classifier : 84.2%
Logistic Regression : 94.7%

(2) **recall**
Decision tree :  89% 
random forest : 100%
svm : 86%
sgd classifier : 78%
Logistic Regression : 93%


유방암 진단의 경우 양성인데 음성을 판단하는 경우가 적어야 하기 때문에(양성을 놓치면 안되기 때문에) Recall로 모델을 평가하는 것이 좋다.
accuracy와 recall이 100%로 모두 높은 random forest 모델을 사용하는 것이 가장 좋다고 판단된다.


---



2022-01-06

---